mqtt/policy/src/core/builder.rs

```diff
@@ -0,0 +1,706 @@
+
+    static_rules.merge(static_ids);
+    variable_rules.merge(variable_ids);
+}
+
+fn process_identities(statement: &Statement20201030) -> (Identities, Identities) {
+    let mut static_ids = Identities::new();
+    let mut variable_ids = Identities::new();
+    for identity in &statement.identities {
+        let (static_ops, variable_ops) = process_operations(&statement);
+
+        if is_variable_rule(identity) {
+            // if current identity has substitutions,
+            // then the whole operation subtree need
+            // to be cloned into substitutions tree.
+            let mut all = static_ops.clone();
+            all.merge(variable_ops);
+            variable_ids.insert(identity, all);
+        } else {
+            // else, divide operations and operation substitutions
+            // between identities and identity substitutions.
+            static_ids.insert(identity, static_ops);
+            variable_ids.insert(identity, variable_ops);
+        }
+    }
+
+    (static_ids, variable_ids)
+}
+
+fn process_operations(statement: &Statement20201030) -> (Operations, Operations) {
+    let mut static_ops = Operations::new();
+    let mut variable_ops = Operations::new();
+    for operation in &statement.operations {
+        let (static_res, variable_res) = process_resources(&statement);
+
+        if is_variable_rule(operation) {
+            // if current operation has variables,
+            // then the whole resource subtree need
+            // to be cloned into variables tree.
+            let mut all = static_res.clone();
+            all.merge(variable_res);
+            variable_ops.insert(operation, all);
+        } else {
+            // else, divide static resources and variable resources
+            // between static operations and variable operation.
+            static_ops.insert(operation, static_res);
+            variable_ops.insert(operation, variable_res);
+        }
+    }
+
+    (static_ops, variable_ops)
+}
+
+fn process_resources(statement: &Statement20201030) -> (Resources, Resources) {
+    let mut static_res = Resources::new();
+    let mut variable_res = Resources::new();
+    if statement.resources.is_empty() {
+        static_res.insert("", statement.into());
+    }
+
+    for resource in &statement.resources {
+        // split resources into two static or variable rules:
+        let map = if is_variable_rule(resource) {
+            &mut variable_res
+        } else {
+            &mut static_res
+        };
+
+        map.insert(resource, statement.into());
+    }
+
+    (static_res, variable_res)
+}
+
+fn is_variable_rule(value: &str) -> bool {
+    lazy_static! {
+        static ref VAR_PATTERN: Regex =
+            Regex::new(r#"\{\{[^\{\}]+\}\}"#).expect("failed to create a Regex from pattern");
+    }
+    VAR_PATTERN.is_match(value)
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct PolicyDefinition20201030 {
```

Comment:
How often do we see this naming scheme changing? 
Are Statement/Effect/PolicyDefinition tied together?

Reply:
>How often do we see this naming scheme updating?

I'm not sure how often that would be, but we need to think about versioning.

>Are Statement/Effect/PolicyDefinition tied together?

yes, in current policy definition schema.